lib/semantics/check-do.cc

```diff
@@ -583,48 +574,68 @@ class DoContext {
     }
   }
 
-  // check constraints [C1121 .. C1130]
-  void EnforceConcurrentLoopControl(
-      const parser::LoopControl::Concurrent &concurrent,
-      const parser::Block &block) const {
-
-    auto &header{std::get<parser::ConcurrentHeader>(concurrent.t)};
+  // C1123, concurrent limit or step expressions can't reference index-names
+  void CheckConcurrentHeader(const parser::ConcurrentHeader &header) const {
     auto &controls{std::get<std::list<parser::ConcurrentControl>>(header.t)};
-    SymbolContainer indexNames;
-    for (auto &c : controls) {
-      auto &indexName{std::get<parser::Name>(c.t)};
+    SymbolSet indexNames;
+    for (const auto &c : controls) {
+      const auto &indexName{std::get<parser::Name>(c.t)};
       if (indexName.symbol) {
         indexNames.insert(indexName.symbol);
       }
     }
     if (!indexNames.empty()) {
-      for (auto &c : controls) {
-        // C1123
+      for (const auto &c : controls) {
         HasNoReferences(indexNames, std::get<1>(c.t));
         HasNoReferences(indexNames, std::get<2>(c.t));
-        if (auto &expression{
+        if (const auto &expr{
                 std::get<std::optional<parser::ScalarIntExpr>>(c.t)}) {
-          HasNoReferences(indexNames, *expression);
+          HasNoReferences(indexNames, *expr);
         }
       }
     }
+  }
 
-    auto &mask{std::get<std::optional<parser::ScalarLogicalExpr>>(header.t)};
-    if (mask.has_value()) {
-      CheckMaskIsPure(*mask);
-    }
-    auto &localitySpecs{
+  void CheckLocalitySpecs(const parser::LoopControl::Concurrent &concurrent,
+      const parser::Block &block) const {
+    const auto &header{std::get<parser::ConcurrentHeader>(concurrent.t)};
+    const auto &controls{
+        std::get<std::list<parser::ConcurrentControl>>(header.t)};
```

Comment:
Is it safe to assume that `controls` will not be empty?

Reply:
Yes.  The associated element of the grammar is a ```concurrent-control-list```.  Section 4.1.3 of the standard in R401 states ```xyz-list is xyz [ , xyz ] ...```.  So if there's not at least one ```concurrent-control```, there will be a parsing error.